notebook/services/kernels/handlers.py

```diff
@@ -461,7 +461,32 @@ def _send_status_message(self, status):
         msg['channel'] = 'iopub'
         self.write_message(json.dumps(msg, default=date_default))
 
-    def on_kernel_restarted(self):
+    def on_kernel_restarted(self, **kwargs):
+        if 'newports' in kwargs:
+            self.log.debug("Kernel restarted with new ports")
+            newports = kwargs['newports']
+        else:
+            self.log.debug('newports parameter is not defined, setting default to False')
+            newports = False
+        logging.info("Restarting kernel with new ports: {}".format(newports))
+        if newports:
+            # Stream should be closed before reconnecting to the new ports
+            for channel, stream in self.channels.items():
+                if stream is not None and not stream.closed():
+                    stream.on_recv(None)
+                    stream.close()
+            # Connect to the stream with new ports
+            try:
+                self.create_stream()
+            except web.HTTPError as e:
```

Comment:
I can see this is copied from the `open()` method, but I don't think `create_stream` can actually raise `HTTPError`. Maybe it could at some point in the past. @minrk any ideas?

Reply:
I think you're right. My guess is there used to be code in there where we were retrieving the kernel, which could raise 404, etc. Or it was just a mistake and this should have been Exception all along, which is what it should be now.